RFS/src/main/java/com/rfs/common/DocumentReindexer.java

```diff
@@ -9,33 +11,51 @@
 import org.opensearch.migrations.reindexer.tracing.IDocumentMigrationContexts;
 
 import lombok.RequiredArgsConstructor;
+import reactor.core.publisher.BufferOverflowStrategy;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.function.Tuple2;
 
 @RequiredArgsConstructor
 public class DocumentReindexer {
     private static final Logger logger = LogManager.getLogger(DocumentReindexer.class);
     protected final OpenSearchClient client;
     private final int numDocsPerBulkRequest;
+    private final long numBytesPerBulkRequest;
     private final int maxConcurrentRequests;
+    private final double maxRequestsPerSecond;
 
     public Mono<Void> reindex(
         String indexName,
         Flux<Document> documentStream,
         IDocumentMigrationContexts.IDocumentReindexContext context
     ) {
-
-        return documentStream.map(this::convertDocumentToBulkSection)  // Convert each Document to part of a bulk
-                                                                       // operation
-            .buffer(numDocsPerBulkRequest) // Collect until you hit the batch size
-            .doOnNext(bulk -> logger.info("{} documents in current bulk request", bulk.size()))
-            .map(this::convertToBulkRequestBody)  // Assemble the bulk request body from the parts
+        final int requestBuffer = Math.min(Math.max((int) maxRequestsPerSecond * 30, 10), 100); // After a pause/slowdown, allow up to 30 seconds of bursting or 100 requests
+        return Flux.interval(Duration.ofMillis((long) (1000 / maxRequestsPerSecond)), Schedulers.single())
+            .onBackpressureBuffer(requestBuffer, BufferOverflowStrategy.DROP_OLDEST)  // Drop ticks on backpressure, note requests are not dropped
+            .zipWith(documentStream
+                .publishOn(Schedulers.single())
+                .map(this::convertDocumentToBulkSection)  // Convert each Document to part of a bulk operation
+                .bufferWhile(bufferPredicate(numDocsPerBulkRequest, numBytesPerBulkRequest)) // Collect until you hit the batch size or max size
```

Comment:
bufferWhile really means bufferWhileTrueAndFlushAndRestartWhenFalse?

Reply:
Yes, i've since changed this to bufferUntil but similar process